### PR TITLE
apiextensions-apiserver: Fix decoding of DeleteOptions.

### LINKED ISSUE #46736: TPR objects still exist after deleting namespaces
**BUG REPORT**

**Kubernetes version** (use `kubectl version`): v1.6.2

**Environment**:
- **Cloud provider or hardware configuration**: GKE

**What happened**:
- Create TPR "MyResource"
- Create namespace "soak-test"
- Create a "MyResource" "a" under "soak-test":

```
$ cat my-resource-a.yaml
apiVersion: "example.com/v1"
kind: "MyResource"
metadata:
  name: "a"
...
$ kubectl --namespace soak-test create -f my-resource-a.yaml
$ kubectl --namespace soak-test get MyResource
NAME           LABELS    DATA
a           ...
```
- Delete namespace "soak-test"

```
$ kubectl delete ns soak-test
```
- wait namespace to be deleted

MyResource "a" still exists after ns is deleted.
```
$ kubectl --namespace soak-test get MyResource
NAMESPACE   NAME                          KIND
soak-test   a
```

### LINKED ISSUE #47072: Custom Resources prevent namespace deletion
After #46765 which tries to fix #46736, a namespace containing Custom Resources cannot be deleted (it will be stuck in Terminating).

The error on the apiserver side is:

```
status.go:62] apiserver received an error that is not an metav1.Status: Object 'Kind' is missing in '{"orphanDependents":false}
```

According to the stack trace, the error comes from here:

https://github.com/kubernetes/kubernetes/blob/0cff839/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go#L1102

cc @deads2k

### PR DESCRIPTION
**Pull request by enisoc**
Fixes #47072 by making apiextensions-apiserver capable of decoding unversioned DeleteOptions, rather than only handling Unstructured objects (i.e. Custom Resources).

This also closes #46736 and #37554 since the added regression test works for TPR as well.

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -390,9 +429,14 @@ func (crEncoder) Encode(obj runtime.Object, w io.Writer) error {
...
 }
 
-type UnstructuredCreator struct{}
+type unstructuredCreator struct {
+	delegate runtime.ObjectCreater
+}
 
-func (UnstructuredCreator) New(kind schema.GroupVersionKind) (runtime.Object, error) {
+func (c unstructuredCreator) New(kind schema.GroupVersionKind) (runtime.Object, error) {
+	if obj, err := c.delegate.New(kind); err == nil {
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Doesn't this mean that certain types become invalid targets for CRDS?  `ListOptions`, `ExportOptions`, etc?  Not that I object, I just want to be sure I understand the limitation.

**Comment by deads2k** [comment]
@enisoc prepare to be a target for everyone wondering about why their scheme isn't working :)

/lgtm

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (enisoc) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Do you mean a CRD with Kind `ListOptions` will break? I hadn't thought about that to be honest.

Would it help if I remove the delegation for `New()` and only keep it for the methods where we can distinguish whether the caller specifically expects `Unstructured`? The particular `DeleteOptions` code path we care about here probably doesn't call `New()` anymore now that I've fixed the decoder to not return `Unstructured`.

**Pull request merged**